File: stress/stress/network.py
"""SP-4: Network Jitter — latency variance + packet loss simulation."""
from __future__ import annotations

import random
from dataclasses import dataclass

from stress.stress.base import Stressor, StressorConfig


@dataclass(frozen=True)
class NetworkConfig(StressorConfig):
    """SP-4 configuration."""
    baseline_latency_ms: float = 50.0
    jitter_pct: float = 50.0  # Variance as percentage of baseline
    packet_loss_probability: float = 0.05


class NetworkStressor(Stressor):
    """Network jitter and packet loss simulation.

    Applied to workloads with communication dependencies (W2-A external calls,
    W3-A inter-node messaging).
    """

    def __init__(self, config: NetworkConfig):
        super().__init__(config)
        self._cfg: NetworkConfig = config

    def reset(self) -> None:
        self._rng = random.Random(self._config.seed)
        self._active = False
        self._start_time = None

    def get_latency_ms(self, t: float) -> float:
        """Get simulated network latency at time t.

        Returns baseline + jitter drawn from normal distribution,
        clipped to [0, 2 * baseline].
        """
        if not self._active or not self._cfg.enabled:
            return 0.0

        jitter_std = self._cfg.baseline_latency_ms * (self._cfg.jitter_pct / 100.0)
        latency = self._rng.gauss(self._cfg.baseline_latency_ms, jitter_std)
        return max(0.0, min(latency, 2.0 * self._cfg.baseline_latency_ms))

    def is_packet_lost(self, t: float) -> bool:
        """Check if a packet is lost at time t (Bernoulli trial)."""
        if not self._active or not self._cfg.enabled:
            return False
        return self._rng.random() < self._cfg.packet_loss_probability

File: stress-ref/src/stress/network.rs
use rand::Rng;
use rand::SeedableRng;
use rand_chacha::ChaCha8Rng;

/// SP-4: Network Jitter — latency variance + packet loss.
pub struct NetworkStressor {
    enabled: bool,
    baseline_latency_ms: f64,
    jitter_pct: f64,
    packet_loss_probability: f64,
    rng: ChaCha8Rng,
    active: bool,
}

impl NetworkStressor {
    pub fn new(
        enabled: bool,
        seed: u64,
        baseline_latency_ms: f64,
        jitter_pct: f64,
        packet_loss_probability: f64,
    ) -> Self {
        Self {
            enabled,
            baseline_latency_ms,
            jitter_pct,
            packet_loss_probability,
            rng: ChaCha8Rng::seed_from_u64(seed),
            active: false,
        }
    }

    pub fn start(&mut self, _t: f64) {
        self.active = true;
    }

    pub fn stop(&mut self) {
        self.active = false;
    }

    pub fn enabled(&self) -> bool {
        self.enabled
    }

    pub fn get_latency_ms(&mut self, _t: f64) -> f64 {
        if !self.active || !self.enabled {
            return 0.0;
        }
        let jitter_std = self.baseline_latency_ms * (self.jitter_pct / 100.0);
        // Box-Muller transform for normal distribution
        let u1: f64 = self.rng.gen::<f64>().max(1e-10);
        let u2: f64 = self.rng.gen();
        let z = (-2.0 * u1.ln()).sqrt() * (2.0 * std::f64::consts::PI * u2).cos();
        let latency = self.baseline_latency_ms + jitter_std * z;
        latency.clamp(0.0, 2.0 * self.baseline_latency_ms)
    }

    pub fn is_packet_lost(&mut self, _t: f64) -> bool {
        if !self.active || !self.enabled {
            return false;
        }
        self.rng.gen::<f64>() < self.packet_loss_probability
    }
}
